adr/ADR-43.md

```diff
@@ -0,0 +1,87 @@
+# JetStream Per-Message TTL
+
+| Metadata | Value                     |
+|----------|---------------------------|
+| Date     | 2024-07-11                |
+| Author   | @ripienaar                |
+| Status   | Approved                  |
+| Tags     | jetstream, client, server |
+
+## Context and motivation
+
+Streams support a one-size-fits-all approach to message TTL based on the MaxAge setting. This causes any message in the 
+Stream to expire at that age.
+
+There are numerous uses for a per-message version of this limit, some listed below:
+
+ * KV tombstones are a problem in that they forever clog up the buckets with noise, these could have a TTL to make them expire once not useful anymore
+ * Server-applied limits can result in tombstones with a short per message TTL so that consumers can be notified of limits being processed. Useful in KV watch scenarios being notified about TTL removals
+ * A stream may have a general MaxAge but some messages may have infinite retention, think a schema or type hints in a KV bucket that is forever while general keys have TTLs
+ * TODO: user use cases
+
+Related issues [#3268](https://github.com/nats-io/nats-server/issues/3268)
+
+## Per-Message TTL
+
+We will allow a message to supply a TTL as a duration in a header called `Nats-TTL` followed by the duration as 
+nanoseconds, the sever may not supper nano second level precision and clients can choose to expose TTL as
+seconds or nanoseconds.
+
+The duration will be used by the server to calculate the deadline for removing the message based on its Stream 
+timestamp and the stated duration.
+
+The TTL may not exceed the Stream MaxAge to give administrators a hard cap for the purposes of managing resources.
+
+A TTL of zero will be ignored, any other unparsable value will result in a error reported in the Pub Ack and the message
+being discarded.
+
+Setting the header `Nats-No-Expire` to `1` will result in a message that will never be expired.
+
+## Limit Tombstones
+
+Several scenarios for server-created tombstones can be imagined, the most often requested one though is when MaxAge
+removes last value (ie. the current value) for a Key.
+
+In this case when the server removes a message and the message is the last in the subject it would place a message 
+with a TTL matching the Stream configuration value.  The following headers would be placed:
+
+```
+Nats-Applied-Limit: MaxAge
+Nats-TTL: 1000000000
+```
+
+The `Nats-Limit-Applied` field is there to support future expansion of this feature.
+
+This behaviour is off by default unless opted in on the Stream Configuration.
+
+## Publish Acknowledgements
+
+We could optionally extend the `PubAck` as follows:
+
+```golang
+type PubAck struct {
+	MsgTTL    uint64 `json:"msg_ttl,omitempty"`
```

Comment:
There isnt a guarantee the server will expire the message at any exact deadline time stamp so I was reluctant to put in a timestamp here.

Reply:
Also no guarantee about the server time vs client time, so unless the puback also included the server timestamp it still would be an unknown.  I've wanted to include the ts in the ack but suspect it will be too many bytes for not a lot of utilty.

Reply:
possibly it should be a concrete timestamp in the future. If the data is invalid after a certain amount of time, it is certainly invalid if the server was not able to remove it because it was down, but came up.